design-documents/node-static-content-deploy.md

```diff
@@ -0,0 +1,95 @@
+    - **Quick**: Duplicates a ton of files, which makes cloud deployments slower
+    - **Standard**: Does way more work than it needs to, and duplicates tons of files
+    - **Compact**: Basically re-implements symlinks in the application layer, ships 150kb+ of extra JS to the front-end
+1. When something fails, it's very difficult to figure out why
+1. Lack of reliable cache + incremental builds (most people typically run it from scratch)
+1. Unable to introduce modern FE tooling (because it all expects a `node.js` runtime)
+1. Lacking granular reporting (unclear which parts of the process cause spikes in deployment times)
+
+## Why rewrite Static Content Deployment?
+
+There are 3 primary motivators for re-writing the Static Content Deployment tool
+
+### Performance
+
+The performance of `bin/magento setup:static-content:deploy` is _not great_, which leads to developers being unhappy about doing front-end work.
+
+It is also a very significant bottleneck for Magento Cloud deployments.
+
+### Access to Modern Tooling
+The majority of modern front-end tooling is written in JavaScript and executed using `node.js`. By moving our front-end compilation tool to JavaScript, we're able to leverage a large ecosystem of open-source tooling that cannot easily be interfaced with from PHP.
+
+A great example of this is our usage of LESS. The only [official compiler for the LESS language](http://lesscss.org/usage/) is written in JavaScript. Because of this, Magento uses a 3rd party (incomplete) implementation of the LESS compiler in PHP. However, because of the slowness associated with it, most developers prefer to use Magento's grunt setup _during local development_. The differences between these 2 compilers leads to [very unfortunate bugs](https://github.com/magento/magento2/issues/7231) when a production deployment is done.
+
+### Accessible for Front-End Devs
+When building something new, it's important to choose the right programming language for the job.
+
+Often, the "right" language is based on the skillset of those that will be maintaining the tool. As an example, the Magento CLI is written using PHP. PHP is not a language designed with the intent of being used to write CLIs, but it is capable of it. Magento chose to write `bin/magento` in PHP because it makes contribution accessible to those using the tool.
+
+A re-write in JavaScript means that the code for static content deployment can be maintained by front-end developers, who spend most of their time in Magento working on JavaScript/HTML/CSS.
+
+### What opportunities open up with a re-write of SCD using node.js?
+- Compilation of modern JS with [Babel](https://babeljs.io/)
+- Automatic live reloading of CSS during development
```

Comment:
Could you clarify what do you mean? Now, in the development mode, the css/js will be generated in runtime.

Reply:
Example:

You are on the checkout page. You have 3/4 of the checkout form filled out. You now want to tweak some values in CSS.

Currently, your options are:

1. Make the change in DevTools, then copy/paste into your store CSS
2. Make the change on the file system, then refresh the whole page and re-enter all form data

Live reloading injects the CSS changes into the live document - you get to keep your browser state, but use your own editor.

Reply:
Anyone who has worked on a dev server with hot reloading (pretty common among FE devs) would be offended at the non-existence of it. It's really painful to not have it nowadays. We need fast, iterative development cycles and if we don't have live reloading, it slows us down (a lot).

Reply:
Should be there some kind of file watcher? Will it be a separate tool or an existing one?

Reply:
Yep! That would be the idea for development mode (when actively working on front-end tasks). You typically have a persistent process running that keeps a warm cache and does small incremental rebuilds when files change (which typically includes running linters/unit tests/etc).

The [typical file watcher](https://github.com/paulmillr/chokidar) lib used in node tooling supports the native watch mechanisms on most operating systems, but can fall back to polling for mounts (for our friends developing with vagrant).

Reply:
Do you still plan to have some development mode auto processing of missing assets?

From PoV of a agency that is doing things own way - kill it with fire

From an imaginary possition of full stacks, backends and other people who might not know modern frontend tooling - "it doesnt't work after changes, just got a page without sytes and can't click anything" - at least that what I see from questions asked over years about SASS Blank which requires to manually run some tools to get styles.

Reply:
I think we need some solution here. Just like I don't want to setup docker + a bunch of microservices locally, most backend folks won't want to be running a persistent node process.

We'll find a good middle-ground here.